src/routes/examples/main/react.mdx

```diff
@@ -1,3 +1,250 @@
+
+```jsx
+// src/App.js
+
+import { useState, useEffect } from "react"
+
+export default function App() {
+  let [users, setUsers] = useState([])
+
+  useEffect(() => {
+    let fetchUsers = async function() {
+      let res = await fetch("/api/users")
+      let json = await res.json()
+      setUsers(json.users)
+    }
+
+    fetchUsers()
+  }, [])
+
+  return (
+    <ul data-testid="users">
+      {users.map(user => (
+        <li key={user.id} data-testid={`user-${user.id}`}>
+          {user.name}
+        </li>
+      ))}
+    </ul>
+  )
+}
+```
+
+That's it! Our component will now render a list with both of our users, Bob and Alice.
+
+## Using our server in Testing
+
+Next, let's use our Mirage server to test our App component.
+
+> For this next section we're going to use Jest with React Testing Library. If you're unfamiliar with these libraries take a look at their [guides](https://testing-library.com/docs/react-testing-library/intro).
+
+Create a new test for the App component in `src/__tests__/App.test.js`. To get started, copy in the following Jest code.
+
+```js
+// src/__tests__/App.test.js
+
+import React from "react";
+import {
+  render,
+  waitForElement,
+} from "@testing-library/react";
+import App from "../App";
+import { makeServer } from "../server";
+
+let server;
+
+beforeEach(() => {
+  server = makeServer({ environment: "test" });
+});
+
+afterEach(() => {
+  server.shutdown();
+});
+
+it("loads the users from our server", () => {
+  const { getByTestId } = render(<App />);
+
+  await waitForElement(() => getByTestId("user-1"));
+  await waitForElement(() => getByTestId("user-2"));
+
+  expect(getByTestId("user-1")).toHaveTextContent("Bob");
+  expect(getByTestId("user-2")).toHaveTextContent("Alice");
+});
+```
+
+Let's talk about what's going on in this test file.
+
+In the `beforeEach` hook we're creating a new server using the same `makeServer` function we defined earlier. We're initializing our server with a `test` environment, which optimizes Mirage from running inside of tests.
+
+In the `afterEach` hook we're calling `server.shutdown()`, which will destroy our Mirage server. This ensures that every test we write has its own isolated server and no server state is shared between any two tests.
+
+Finally, our test renders our App component and asserts that the names Bob and Alice show up on the page.
+
+While this test covers the data loading functionality of our component there's something missing. When we test components like this, we maximize our confidence in them by testing them in various states. When dealing with a data loading component like our App component, we want to be able to answer questions like "What happens when the server has no users?" or "What does our component render if the backend 500s?"
```

Comment:
Small tweak here?

> ...something missing. Testing components like this will maximize our confidence in them being able to take on various states. However, we also must be able to answer additional questions like: "What happens when the server has no users?", or "What does our component render if the backend 500's?"